### PR TITLE
Integrate costmap in velocity filter

### PR DESCRIPTION
**Pull request by ghost**
... instead of having a separate node

### CODE REVIEW

File: cob_collision_velocity_filter/src/cob_collision_velocity_filter.cpp [hunk 1]
```diff
@@ -52,62 +52,54 @@
...
 
   nh_.param("costmap_obstacle_treshold", costmap_obstacle_treshold_, 50);
 
   // implementation of topics to publish (command for base and list of relevant obstacles)
   topic_pub_command_ = nh_.advertise<geometry_msgs::Twist>("command", 1);
   topic_pub_relevant_obstacles_ = nh_.advertise<nav_msgs::OccupancyGrid>("relevant_obstacles_grid", 1);
 
+  std::string teleop_topic;
+  if(!nh_.hasParam("teleop_topic")) ROS_WARN("Used default parameter for teleop_topic [teleop_twist]");
+  nh_.param("teleop_topic",teleop_topic, std::string("teleop_twist"));
```
Comment (mgruhler) [inline cob_collision_velocity_filter/src/cob_collision_velocity_filter.cpp]: we don't set topics by parameters.
Please remove.

This should be done via remappings in the respective launch files.

File: cob_collision_velocity_filter/CMakeLists.txt [hunk 2]
```diff
@@ -1,7 +1,16 @@
...
+find_package(catkin REQUIRED COMPONENTS
+cob_footprint_observer
+costmap_2d
+dynamic_reconfigure
+geometry_msgs
+nav_msgs
+roscpp
+tf
+visualization_msgs
+)
```
Comment (mgruhler) [inline cob_collision_velocity_filter/CMakeLists.txt:13]: please indent

File: cob_collision_velocity_filter/src/cob_collision_velocity_filter.cpp [hunk 3]
```diff
@@ -128,12 +117,12 @@ CollisionVelocityFilter::CollisionVelocityFilter()
   if(!nh_.hasParam("pot_ctrl_virt_mass")) ROS_WARN("Used default parameter for pot_ctrl_virt_mass [0.8]");
   nh_.param("pot_ctrl_virt_mass", virt_mass_, 0.8);
 
-  //load the robot footprint from the parameter server if its available in the local costmap namespace
-  robot_footprint_ = loadRobotFootprint(local_costmap_nh_);
+  robot_footprint_ = anti_collision_costmap_->getRobotFootprint();
+
   if(robot_footprint_.size() > 4)
-    ROS_WARN("You have set more than 4 points as robot_footprint, cob_collision_velocity_filter can deal only with rectangular footprints so far!");
+  ROS_WARN("You have set more than 4 points as robot_footprint, cob_collision_velocity_filter can deal only with rectangular footprints so far!");
```
Comment (mgruhler) [inline cob_collision_velocity_filter/src/cob_collision_velocity_filter.cpp]: indent

File: cob_collision_velocity_filter/src/cob_collision_velocity_filter.cpp [hunk 4]
```diff
@@ -176,37 +171,13 @@ void CollisionVelocityFilter::joystickVelocityCB(const geometry_msgs::Twist::Con
...
+void CollisionVelocityFilter::getFootprintServiceCB(const ros::TimerEvent& event)
 {
-  cob_footprint_observer::GetFootprint srv = cob_footprint_observer::GetFootprint();
-  // check if service is reachable
-  if (srv_client_get_footprint_.call(srv))
-  {
-    // adjust footprint
-    geometry_msgs::PolygonStamped footprint_poly = srv.response.footprint;
+    ROS_DEBUG("[cob_collision_velocity_filter] Update footprint");
+     // adjust footprint
```
Comment (mgruhler) [inline cob_collision_velocity_filter/src/cob_collision_velocity_filter.cpp]: indentation

**Comment by mgruhler** [comment]
@mig-em looks good

some minor issues:
- please double check indentation (commented some, e.g. in `getFootprintServiceCB`)
- for that matter, please rename this function, as this isn't a service anymore
- remove `srv_client_get_footprint` in header
- remove function `loadRobotFootprint` in header and cpp

File: cob_collision_velocity_filter/src/cob_collision_velocity_filter.cpp [hunk 5]
```diff
@@ -73,41 +68,35 @@ CollisionVelocityFilter::CollisionVelocityFilter()
   topic_pub_relevant_obstacles_ = nh_.advertise<nav_msgs::OccupancyGrid>("relevant_obstacles_grid", 1);
 
   // subscribe to twist-movement of teleop
-  joystick_velocity_sub_ = nh_.subscribe<geometry_msgs::Twist>("teleop_twist", 1, boost::bind(&CollisionVelocityFilter::joystickVelocityCB, this, _1));
-  // subscribe to the costmap to receive inflated cells
-  obstacles_sub_ = nh_.subscribe<nav_msgs::OccupancyGrid>("obstacles", 1, boost::bind(&CollisionVelocityFilter::obstaclesCB, this, _1));
-
-  // create service client
-  srv_client_get_footprint_ = nh_.serviceClient<cob_footprint_observer::GetFootprint>("/get_footprint");
+  joystick_velocity_sub_ = nh_.subscribe<geometry_msgs::Twist>("/base/collision_velocity_filter/command", 10, &CollisionVelocityFilter::joystickVelocityCB, this);
```
Comment (mgruhler) [inline cob_collision_velocity_filter/src/cob_collision_velocity_filter.cpp]: This should be simpler and not a global topic. Please rename to `"command_in"`.

**Comment by ghost** [comment]
Please look at the little fixes commit for the changes, "Code reformat with ROS Style" is ONLY code reformat in eclipse.

**Comment by mgruhler** [comment]
One last thing:

please add a public and a private NodeHandle

``` C++
nh_ = ros::NodeHandle(""); // public
pnh_ = ros::NodeHandle("~"); // private
```

Publishers/Subscribers/etc should be on the public, parameters on the private NodeHandle.
This needs to be reflected in the remappings in the launch file as well: i.e. remove the `~` in front of the topics...

**Comment by ghost** [comment]
I suppose the timer should be on the private NodeHandle, right?

**Comment by mgruhler** [comment]
Good question. As long as it works...

Let's see what the community says: http://answers.ros.org/question/233222/ros-timer-difference-between-types-of-nodehandles/

**Comment by mgruhler** [comment]
CR: ok

**Comment by mgruhler** [comment]
should be tested on robots, though.

**Comment by mgruhler** [comment]
Tested on r@w3-3: ok

@ipa-nhg @ipa-fmw Do you want to test this on cob4 as well?

**Comment by NadiaHG** [comment]
FR: OK
I will test  the functionality tomorrow on cob4

**Comment by mgruhler** [comment]
great. so let's wait for the test results.
@ipa-nhg please remember to do the remap to test with teleop.

**Comment by NadiaHG** [comment]
Test on cob4-5: OK

**Comment by mgruhler** [comment]
@ipa-nhg seeing that ipa320/cob_robots#470 has been merged, which actually requires this one, is there any reason why this hasn't been merged yet?

**Comment by NadiaHG** [comment]
@ipa-mig You are right can be merged.

**Pull request merged**